juju/main.py

```diff
@@ -0,0 +1,164 @@
+#!/usr/bin/env python3
+
+import os
+import sys
+import yaml
+
+# Use modules from the $JUJU_CHARM_DIR/lib directory. An absolute path
+# is used as using a symlink like hooks/install -> ../lib/juju/main.py
+# will result in $JUJU_CHARM_DIR/lib/juju being the first path stored
+# in sys.path during hook execution which is undesirable.
+# This happens because the directory containing a script is added to
+# sys.path by the interpreter which gets resolved if it is a symlink.
+del sys.path[0]
+sys.path.insert(0, os.path.join(os.environ.get("JUJU_CHARM_DIR", ''), 'lib'))
+
+import juju.framework # noqa
+
+from juju.framework import ( # noqa
+    Framework,
+    Event,
+    BoundEvent
+)
+from charm import Charm # noqa
+
+# Reuse the same db file as used by charm-helpers.
+DB_PATH = os.path.join(
+    os.environ.get('CHARM_DIR', ''), '.unit-state.db')
+
+try:
+    from charmhelpers.core.hookenv import log
+except (ImportError, ModuleNotFoundError):
+    log = print
+
+
+def load_metadata():
+    with open('metadata.yaml') as f:
+        metadata = yaml.load(f)
+    return metadata
+
+
+def get_charm_endpoints():
+    """Get a list of endpoints from the charm metadata.
+    """
+    metadata = load_metadata()
+    charm_endpoints = list()
+    for endpoint_type in ['provides', 'requires', 'peers']:
+        endpoints = metadata.get(endpoint_type)
+        if endpoints:
+            charm_endpoints.extend(endpoints)
+    return charm_endpoints
+
+
+def setup_hooks():
+    """Set up hooks for supported events.
+
+    Whether a charm can handle an event or not can be determined by
+    introspecting which events are bound to it.
+    """
+
+    relation_events = list()
+    for endpoint in get_charm_endpoints():
+        relation_events.extend([
+            f'{endpoint}_relation_{t}'
+            for t in ['joined', 'changed', 'departed', 'broken']
+        ])
+
+    log(f'Discovered relation events from charm metadata: {relation_events}')
```

Comment:
This is not logging.. this is fine-grained debugging. It's fine to have it.. we just need to not mix the two, because one has developers/power users as an audience, and the other has system log files as that.